pkg/plugins/storage/azblob/azblob.go

```diff
@@ -0,0 +1,185 @@
+	azureutils "github.com/nitric-dev/membrane/pkg/providers/azure/utils"
+	"github.com/nitric-dev/membrane/pkg/utils"
+)
+
+// AzblobStorageService - Nitric membrane storage plugin implementation for Azure Storage
+type AzblobStorageService struct {
+	client azblob_service_iface.AzblobServiceUrlIface
+}
+
+func (a *AzblobStorageService) getBlobUrl(bucket string, key string) azblob_service_iface.AzblobBlockBlobUrlIface {
+	cUrl := a.client.NewContainerURL(bucket)
+	// Get a new blob for the key name
+	return cUrl.NewBlockBlobURL(key)
+}
+
+func (a *AzblobStorageService) Read(bucket string, key string) ([]byte, error) {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Read",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+	// Get the bucket for this bucket name
+	blob := a.getBlobUrl(bucket, key)
+	//// download the blob
+	r, err := blob.Download(
+		context.TODO(),
+		0,
+		azblob.CountToEnd,
+		azblob.BlobAccessConditions{},
+		false,
+		azblob.ClientProvidedKeyOptions{},
+	)
+
+	if err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"Unable to download blob",
+			err,
+		)
+	}
+
+	// TODO: Configure retries
+	data := r.Body(azblob.RetryReaderOptions{MaxRetryRequests: 20})
+
+	return ioutil.ReadAll(data)
+}
+
+func (a *AzblobStorageService) Write(bucket string, key string, object []byte) error {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Write",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+
+	blob := a.getBlobUrl(bucket, key)
+
+	if _, err := blob.Upload(
+		context.TODO(),
+		bytes.NewReader(object),
+		azblob.BlobHTTPHeaders{},
+		azblob.Metadata{},
+		azblob.BlobAccessConditions{},
+		azblob.DefaultAccessTier,
+		nil,
+		azblob.ClientProvidedKeyOptions{},
+	); err != nil {
+		return newErr(
+			codes.Internal,
+			"Unable to write blob data",
+			err,
+		)
+	}
+
+	return nil
+}
+
+func (a *AzblobStorageService) Delete(bucket string, key string) error {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Delete",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+
+	// Get the bucket for this bucket name
+	blob := a.getBlobUrl(bucket, key)
+
+	if _, err := blob.Delete(
+		context.TODO(),
+		azblob.DeleteSnapshotsOptionInclude,
+		azblob.BlobAccessConditions{},
+	); err != nil {
+		return newErr(
+			codes.Internal,
+			"Unable to delete blob",
+			err,
+		)
+	}
+
+	return nil
+}
+
+const expiryBuffer = 2 * time.Minute
+
+func tokenRefresherFromSpt(spt *adal.ServicePrincipalToken) azblob.TokenRefresher {
+	return func(credential azblob.TokenCredential) time.Duration {
+		if err := spt.Refresh(); err != nil {
+			fmt.Println("Error refreshing token: ", err)
+		} else {
+			tkn := spt.Token()
+			credential.SetToken(tkn.AccessToken)
+
+			return tkn.Expires().Sub(time.Now().Add(expiryBuffer))
+		}
+
+		// Mark the token as already expired
+		return time.Duration(0)
+	}
```

Comment:
Is there any scope for having some common Azure token refresh code?

Reply:
There is but need to examine the interface chain for the structs that implement these tokens for both azqueue and azblob to see where there is commonality, which can be done in a later PR.